### PR TITLE
core/deadline: ignore deadlined duties

### PR DESCRIPTION
**Pull request by xenowits**
- Modify `Deadliner` to ignore deadlined duties.
- Modify `Add(duty)` to return true if a duty is added successfully.
- Fix deadline tests.

category: refactor
ticket: #814

### CODE REVIEW

File: core/deadline.go
Comment (dB2510) [inline core/deadline.go:51]: we decided to return bool from Add function. Not adding a new method. remove this

File: core/deadline.go
Comment (dB2510) [inline core/deadline.go:57]: Add function will return the response of this channel. Modified interface:
```
// Deadliner provides duty Deadline functionality. The C method isn’t thread safe and
// may only be used by a single goroutine. So, multiple instances are required
// for different components and use cases.
type Deadliner interface {
	// Add adds a duty to be notified of the Deadline via C.
	// Note that duties will be deduplicated and only a single duty will be provided via C.
        // It also expects a boolean response which signals if added duty is a expired one or not.
	Add(duty Duty) bool

	// C returns the same read channel every time and contains deadlined duties.
	// It should only be called by a single goroutine.
	C() <-chan Duty
}
```

File: core/deadline.go
Comment (xenowits) [reply core/deadline.go]: done!

File: core/deadline.go
Comment (dB2510) [inline core/deadline.go:98]: replace return with continue with return. return will exit from this goroutine while we still want to add duties

File: core/deadline.go
Comment (dB2510) [inline core/deadline.go:131]: suggest:
```
        select {
	case <-d.quit:
		return false
	default:
		d.dutyChan <- duty
		return <-d.addedChan
	}
```

File: core/deadline_test.go
Comment (dB2510) [inline core/deadline_test.go:39]:
```suggestion
	deadlineFuncProvider := func(duty core.Duty) time.Time {
			if duty.Type == core.DutyExit {
				// Do not timeout exit duties.
				return time.Date(9999, 1, 1, 0, 0, 0, 0, time.UTC)
			}

			return time.Now().Add(time.Millisecond * time.Duration(duty.Slot))
```

Comment (dB2510) [reply core/deadline_test.go:39]: can add custom deadlineFunc which will not depend of any time assumptions and deadline duties in desired order

Comment (dB2510) [reply core/deadline_test.go:39]: Also add tests to ignore expired duties

Comment (xenowits) [reply core/deadline_test.go:39]: > can add custom deadlineFunc which will not depend of any time assumptions and deadline duties in desired order

yes, but i don't think it's required in this test as duties are received in the same order as the order in which they are added

File: core/deadline.go
Comment (corverroos) [inline core/deadline.go:41]: Suggest: `Add returns true if the duty was added for future deadline scheduling. It is idempotent and returns true if the duty was previously added and still awaits deadline scheduling. It returns false if the duty has already expired and cannot therefore be added for scheduling.`

File: core/deadline.go
Comment (corverroos) [inline core/deadline.go:53]: This isn't going to work. You cannot guarantee which reader will receive the message if multiple are listening. Follow the. pattern in `aggsigdb.MemDB.commands.writeCommand`. but rename to:

```
type deadlineInput struct {
  Duty core.Duty
  Success chan <- bool 
}

// Note we can rename and unexport deadliner implementation since we only ever export the interface.
type deadliner struct {
  input chan deadlineInput
  ...
}
```

File: core/deadline_test.go
Comment (xenowits) [reply core/deadline_test.go:39]: > Also add tests to assert ignored expired duties

yes, we do need to also add some expired duties to test that flow, i'm adding that

File: core/deadline.go
Comment (dB2510) [reply core/deadline.go:53]: Based on this i would suggest:
```
func (d *Deadline) Add(duty Duty) {
	select {
	case <-d.quit:
		return
	default:
               input := deadlineInput{duty: duty, success: make(chan bool)}
               d.input <- input
               return <-input.success
	}
}
```

Comment (corverroos) [reply core/deadline.go:53]: Please look closely at aggsigdb.MemDB.commands.writeCommand, the above will not work

Comment (dB2510) [reply core/deadline.go:54]:
```
func (d *Deadline) Add(duty Duty) bool {
        response := make(chan bool)
	select {
	case <-d.quit:
		return false
	default:
               d.input <- deadlineInput{duty: duty, success: response}
	}

        select {
	case <-d.quit:
		return false
	default:
                return <-response
	}
}
```

Comment (corverroos) [reply core/deadline.go:54]: Question, why not used buffered channel `response := make(chan bool, 1)` like aggsigdb?

Comment (dB2510) [reply core/deadline.go:54]: Buffered channel would be useful to unblock deadliner's central goroutine to not wait for the reader of this channel. In case if add goroutine is exited, central go routine should also be exited not to wait for the reader to read response to add.

We should use buffered channel 👍

Comment (corverroos) [reply core/deadline.go:54]: I think either is actually fine in this specific use-case. "Unbuffered" handover is fine since Add goroutine will always already be listening. AggSigDB uses a buffered channel since its Store goroutine may exit early when its context is closed so might not be listening at all. In this case, `Add` doesn't have a context so it cannot exit early so buffered channel isn't required.

File: core/deadline_test.go
Comment (xenowits) [inline core/deadline_test.go:54]: will be adding a `sync.WaitGroup` to wait for these goroutines to finish adding their duties.

Comment (xenowits) [reply core/deadline_test.go:58]: Done!

File: core/deadline.go
Comment (corverroos) [inline core/deadline.go]: please add `*testing.T` after ctx

File: core/deadline_test.go
Comment (xenowits) [inline core/deadline_test.go:81]: different duties are added in random order

File: core/deadline_test.go
Comment (corverroos) [inline core/deadline_test.go:53]: move wg.Add(1) into `addDuties`, since then it adds and remove itself

File: core/deadline_test.go
Comment (corverroos) [inline core/deadline_test.go:62]: why is this required?

File: core/tracker/tracker.go
Comment (corverroos) [inline core/tracker/tracker.go:89]: avoid nesting `if !t.deadliner.Add(e.duty) { // Ignore expired duties; continue }`

**Review by corverroos — APPROVED** [review]

File: core/deadline_test.go
Comment (xenowits) [reply core/deadline_test.go]: To initialize the fake timer

**Pull request merged**
